typed/union_test.go

```diff
@@ -0,0 +1,224 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package typed_test
+
+import (
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/typed"
+)
+
+var unionParser = func() typed.ParseableType {
+	parser, err := typed.NewParser(`types:
+- name: union
+  struct:
+    fields:
+    - name: discriminator
+      type:
+        scalar: string
+    - name: a
+      type:
+        scalar: numeric
+    - name: b
+      type:
+        scalar: numeric
+    - name: c
+      type:
+        scalar: numeric
+    - name: one
+      type:
+        scalar: numeric
+    - name: two
+      type:
+        scalar: numeric
+    - name: three
+      type:
+        scalar: numeric
+    unions:
+      - discriminator: discriminator
+        fields: [a, b, c]
+      - fields: [one, two, three]`)
+	if err != nil {
+		panic(err)
+	}
+	return parser.Type("union")
+}()
+
+func TestNormalizeUnions(t *testing.T) {
+	tests := []struct {
+		name string
+		old  typed.YAMLObject
+		new  typed.YAMLObject
+		out  typed.YAMLObject
+	}{
+		{
+			name: "nothing changed, nothing changes",
+			old:  `{"a": 1}`,
+			new:  `{"a": 1}`,
+			out:  `{"a": 1}`,
+		},
+		{
+			name: "proper union update, setting discriminator",
+			old:  `{"a": 1}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "proper union update, no discriminator",
+			old:  `{"one": 1}`,
+			new:  `{"two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "proper union update from not-set, setting discriminator",
+			old:  `{}`,
+			new:  `{"b": 1}`,
+			out:  `{"b": 1, "discriminator": "b"}`,
+		},
+		{
+			name: "proper union update from not-set, no discriminator",
+			old:  `{}`,
+			new:  `{"two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "remove union, with discriminator",
+			old:  `{"a": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union and discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "remove union, not discriminator",
+			old:  `{"a": 1, "discriminator": "a"}`,
+			new:  `{"discriminator": "a"}`,
+			out:  `{"discriminator": "a"}`,
+		},
+		{
+			name: "remove union, no discriminator",
+			old:  `{"two": 1}`,
+			new:  `{}`,
+			out:  `{}`,
+		},
+		{
+			name: "dumb client update, no discriminator",
+			old:  `{"one": 1}`,
+			new:  `{"one": 2, "two": 1}`,
+			out:  `{"two": 1}`,
+		},
+		{
+			name: "dumb client update, sets discriminator",
+			old:  `{"a": 1}`,
+			new:  `{"a": 2, "b": 1}`,
```

Comment:
I think this should have triggered an error; a dumb client needn't clear a, but it still shouldn't change a and also add b. Thoughts?

Reply:
I really really didn't want to be in the business of comparing deeply nested sub-structures, where fields can have been dropped because of Golang. My goal was very clear: compare the list of fields set, not their content :-/

Reply:
I'm on the fence for update, but for apply--consider what happens the next time they apply if `{"a":2,"b":1}` is what is in their local file. I think we should attempt to prevent them from having a local file that sets multiple fields.

I guess think update should prevent this, too, except in cases where that is adding an additional validation constraint--for those we need validation latching.

Reply:
I thought about this, but we can't detect that once we have merged the object, so I was planning on adding another validation step just for that before we merge.